src/charm.py

```diff
@@ -18,3 +24,56 @@ def __init__(self, *args):
             args: Arguments passed to the CharmBase parent constructor.
         """
         super().__init__(*args)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.saml_relation_created, self._on_relation_created)
+
+    def _on_config_changed(self, _) -> None:
+        """Handle changes in configuration."""
+        self.unit.status = ops.MaintenanceStatus("Configuring charm")
+        try:
+            self._charm_state = CharmState.from_charm(charm=self)
+            self._saml_integrator = SamlIntegrator(charm_state=self._charm_state)
+        except CharmConfigInvalidError as exc:
+            self.model.unit.status = ops.BlockedStatus(exc.msg)
+            return
+        if self.model.unit.is_leader():
+            for relation in self.model.relations["saml"]:
+                relation.data[self.model.app].update(self.dump_saml_data())
+        self.unit.status = ops.ActiveStatus()
+
+    def dump_saml_data(self) -> Dict[str, str]:
```

Comment:
Should this logic either be in the state or a separate module?

Reply:
like the SAML module

Reply:
Same as above

Reply:
Generally speaking, the charm should only have things on it that interact with juju or pebble in some way, and then rely on individual modules to do the rest of the logic. The logic in this function doesn't need the charm context - it might need the `self._saml_integrator` data to be able to convert it to a form that makes sense for the consumer

Reply:
I'd say it does. The resulting dictionary is a result of the data format supported by juju relations, hance, is charm dependant

Reply:
Same comment, this should probably be managed in the Provider library

Reply:
Indeed. Let's pick up the discussion when the lib is added if you are ok with that. I'm closing the discussion